### PR TITLE
[Ascend] torch_npu.npu_mrope for MRotaryEmbedding

### PR DESCRIPTION
**Pull request by Makcum888e**
## Motivation

use of `torch_npu` implementation of `mrope` instead of torch native implementation with many small operations can give a significant increase of performance

## Modifications

`forward` renamed to `forward_native`
overridden functions `forward_cpu` and `forward_cuda` to not use it from base class
overridden function `forward_npu` with torch_npu implementation

## Accuracy Tests
tested on `Qwen3-VL-4B-Instruct`

`GSM8k`:
before
Accuracy: 0.729
Invalid: 0.122
Latency: 216.923 s
Output throughput: 961.137 token/s

after
Accuracy: 0.728
Invalid: 0.124
Latency: 178.438 s
Output throughput: 1155.518 token/s

`MMMU`:
before
Benchmark time: 221.995
Overall accuracy: 0.499

after:
Benchmark time: 208.065
Overall accuracy: 0.496

## Benchmarking and Profiling

`bench_serving` num prompts 64 random input len 2048 random output len 2048
before
Mean TTFT (ms): 3667.45
Mean ITL (ms): 49.65

after
Mean TTFT (ms): 3494.33
Mean ITL (ms): 43.61

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/layers/rotary_embedding.py [hunk 1]
```diff
@@ -1040,7 +1040,7 @@ def __init__(
                 )
 
     @torch.compile(dynamic=True, backend=get_compiler_backend())
-    def forward(
+    def forward_native(
```
[the remark this answers is not in the document]
Comment (Makcum888e) [reply python/sglang/srt/layers/rotary_embedding.py]: forward_npu added

File: python/sglang/srt/layers/rotary_embedding.py [hunk 2]
```diff
@@ -1087,6 +1087,22 @@ def forward(
...
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        return self.forward_native(positions, query, key)
+
+    def forward_cuda(
+        self,
+        positions: torch.Tensor,
+        query: torch.Tensor,
+        key: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        return self.forward_native(positions, query, key)
```
[the remark this answers is not in the document]
Comment (Makcum888e) [reply python/sglang/srt/layers/rotary_embedding.py]: currently mRotaryEmbedding expected only such inputs, base class can be refactored in future to make it more general

**Review by ssshinigami — APPROVED** [review]
LGTM

**Review by ssshinigami — APPROVED** [review]
LGTM

File: python/sglang/srt/layers/rotary_embedding.py [hunk 3]
```diff
@@ -1489,6 +1490,31 @@ def _forward_triton(
         key = torch.cat((key_rot, key_pass), dim=-1).reshape(key_shape)
         return query, key
 
+    def _forward_npu(
+        self,
+        positions: torch.Tensor,
+        query: torch.Tensor,
+        key: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if query.shape[1] > 4096:
```
Comment (ping1jing2) [inline python/sglang/srt/layers/rotary_embedding.py:1523]: Kindly insert some comments to explain what the magic number `4096` stands for

File: python/sglang/srt/layers/rotary_embedding.py [hunk 4]
```diff
@@ -1435,8 +1435,9 @@ def forward(
             key: [num_tokens, num_kv_heads * head_size]
         """
         assert positions.ndim == 1 or positions.ndim == 2
-
-        if positions.ndim == 2 and self.mrope_section and _is_cuda:
+        if _is_npu:
+            return self._forward_npu(positions, query, key)
+        elif positions.ndim == 2 and self.mrope_section and _is_cuda:
```
Comment (merrymercy) [inline python/sglang/srt/layers/rotary_embedding.py]: can you put the cuda branch as the first one?

Comment (Makcum888e) [reply python/sglang/srt/layers/rotary_embedding.py]: moved

File: python/sglang/srt/layers/rotary_embedding.py (hunk 3 above)
Comment (Makcum888e) [reply python/sglang/srt/layers/rotary_embedding.py:1523]: added

File: python/sglang/srt/layers/rotary_embedding.py [hunk 5]
```diff
@@ -1463,6 +1463,8 @@ def forward(
 
         if positions.ndim == 2 and self.mrope_section and _is_cuda:
             return self._forward_triton(positions, query, key)
+        elif _is_npu:
+            return self._forward_npu(positions, query, key)
```
Comment (ping1jing2) [inline python/sglang/srt/layers/rotary_embedding.py:1466]: same implementation:https://github.com/sgl-project/sgl-kernel-npu/pull/157/files

**Review by ssshinigami — APPROVED** [review]

**Review by hnyls2002 — APPROVED** [review]
